File: Server/ability_validators.py
from constants import SPAWN_CODE, BRIDGE_CODE, D_BRIDGE_CODE, POISON_CODE, NUKE_CODE, CAPITAL_CODE, BURN_CODE, FREEZE_CODE, RAGE_CODE, ZOMBIE_CODE

def no_click(data, player):
    return False


def standard_node_attack(data, player):
    node = data[0]
    return (
        node.owner != player
        and node.owner is not None
        and node.state_name not in ["capital", "mine"]
    )

def my_node(data, player):
    node = data[0]
    return node.owner == player


def dynamic_edge_own_either(data, player):
    edge = data[0]
    return edge.state == "two-way" and (edge.from_node.owner == player)


def capital_logic(data, player):
    node = data[0]
    if (
        node.owner == player
        and node.state_name != "capital"
        and node.full()
    ):
        neighbor_capital = False
        for neighbor in node.neighbors:
            if neighbor.state_name == "capital":
                neighbor_capital = True
                break
        if not neighbor_capital:
            return True
    return False


def standard_port_node(data, player):
    node = data[0]
    return node.owner is not None and node.is_port and node.state_name not in ["mine"]


def unowned_node(data, player):
    node = data[0]
    return node.owner is None and node.state_name == "default"


def new_edge_validator(check_new_edge):
    def new_edge_standard(data, player):
        if len(data) == 1:
            first_node = data[0]
            return first_node.owner == player
        else:
            first_node, second_node = data[0], data[1]
            return first_node.id != second_node.id and check_new_edge(
                first_node.id, second_node.id
            )

    def new_edge_ports(data, player):
        if all([node.port_count > 0 for node in data]):
            return new_edge_standard(data, player)
        return False

    return new_edge_ports


def make_ability_validators(board):
    return {
        SPAWN_CODE: unowned_node,
        BRIDGE_CODE: new_edge_validator(board.check_new_edge),
        D_BRIDGE_CODE: new_edge_validator(board.check_new_edge),
        POISON_CODE: standard_node_attack,
        NUKE_CODE: standard_node_attack,
        CAPITAL_CODE: capital_logic,
        BURN_CODE: standard_port_node, 
        FREEZE_CODE: dynamic_edge_own_either,
        RAGE_CODE: no_click,
        ZOMBIE_CODE: my_node
    }

File: TypeScriptFrontend/src/slav/ability_validators.ts
import { IDItem, Node, Edge, OtherPlayer } from "./Objects";
import { KeyCodes } from "./constants";

type ValidatorFunc = (data: IDItem[]) => boolean;

function noClick(data: IDItem[]): boolean {
    return false;
}

function standardPortNode(data: IDItem[]): boolean {
    const node = data[0] as Node;
    return node.owner !== undefined && node.isPort && node.stateName !== "mine";
}

function capitalValidator(neighbors: (node: Node) => Node[], player: OtherPlayer): ValidatorFunc {
    return function capitalLogic(data: IDItem[]): boolean {
        const node = data[0] as Node;
        if (node.owner === player && node.stateName !== "capital" && node.full) {
            let neighborCapital = false;
            for (const neighbor of neighbors(node)) {
                if (neighbor.stateName === "capital") {
                    neighborCapital = true;
                    break;
                }
            }
            return !neighborCapital;
        }
        return false;
    };
}

function unownedNode(data: IDItem[]): boolean {
    const node = data[0] as Node;
    return node.owner === undefined && node.stateName === "default";
}

function playerValidators(player: OtherPlayer): { [key: string]: ValidatorFunc } {

    // Validator that checks if a node is attackable
    const standardNodeAttack = (data: IDItem[]): boolean => {
        const node = data[0] as Node;  // Type casting to Node for TypeScript
        return node.owner !== player &&
               node.owner !== undefined &&
               node.stateName !== "capital" &&
               node.stateName !== "mine";
    };

    // Validator that checks if a node belongs to the player
    const myNode = (data: IDItem[]): boolean => {
        const node = data[0] as Node;  // Type casting to Node for TypeScript
        return node.owner === player;
    };

    // Validator that checks if either node of a dynamic edge belongs to the player
    const dynamicEdgeOwnEither = (data: IDItem[]): boolean => {
        const edge = data[0] as Edge;  // Type casting to Edge for TypeScript
        return edge.dynamic && (edge.fromNode.owner === player);
    };

    // Return an object mapping codes to their respective validator functions
    return {
        [KeyCodes.POISON_CODE]: standardNodeAttack,
        [KeyCodes.NUKE_CODE]: standardNodeAttack,
        [KeyCodes.FREEZE_CODE]: dynamicEdgeOwnEither,
        [KeyCodes.ZOMBIE_CODE]: myNode,
        [KeyCodes.CANNON_CODE]: myNode,
    };
}

function newEdgeValidator(checkNewEdge: (firstNodeId: number, secondNodeId: number) => boolean, player: OtherPlayer): ValidatorFunc {
    const newEdgeStandard = (data: Node[]): boolean => {
        if (data.length === 1) {
            const firstNode = data[0];
            return firstNode.owner === player;
        } else {
            const firstNode = data[0];
            const secondNode = data[1];
            return firstNode.id !== secondNode.id && checkNewEdge(firstNode.id, secondNode.id);
        }
    };

    return (data: IDItem[]): boolean => {
        const nodes = data as Node[];  // Assert that all data items are Nodes
        return nodes.every(node => node.portCount > 0) && newEdgeStandard(nodes);
    };
}

function makeAbilityValidators(logic: Logic, player: OtherPlayer): {[key: string]: ValidatorFunc} {
    const abilityValidators: {[key: string]: ValidatorFunc} = {
        SPAWN_CODE: unownedNode,
        BRIDGE_CODE: newEdgeValidator(logic.checkNewEdge, player),
        D_BRIDGE_CODE: newEdgeValidator(logic.checkNewEdge, player),
        BURN_CODE: standardPortNode,
        RAGE_CODE: noClick,
        CAPITAL_CODE: capitalValidator(logic.neighbors, player)
    };

    // Merge the validators from `player_validators` into `abilityValidators`
    const playerValidatorsMap = playerValidators(player);
    return {...abilityValidators, ...playerValidatorsMap};
}
